saq/queue/postgres_asyncpg.py

```diff
@@ -0,0 +1,685 @@
+
+        async with self.with_connection(connection) as conn:
+            if job.ttl >= 0:
+                expire_at = int(seconds(now())) + job.ttl if job.ttl > 0 else None
+                await self.update(
+                    job, status=status, expire_at=expire_at, connection=conn
+                )
+            else:
+                await conn.execute(
+                    dedent(f"""
+                    DELETE FROM {self.jobs_table}
+                    WHERE key = $1
+                    """),
+                    key,
+                )
+                await self.notify(job, conn) 
+        await self._release_job(key)
+
+    async def _notify(
+        self,
+        key: str,
+        data: Any | None = None,
+        connection: PoolConnectionProxy | None = None,
+    ) -> None:
+        payload = {"key": key}
+
+        if data is not None:
+            payload["data"] = data
+
+        async with self.with_connection(connection) as conn:
+            await conn.execute(f"NOTIFY \"{self._channel}\", '{json.dumps(payload)}'")
+
+    async def _release_job(self, key: str) -> None:
+        self._releasing.append(key)
+        if self._connection_lock.locked():
+            return
+        async with self._get_dequeue_conn() as conn:
+            txn = conn.transaction()
+            await txn.start()
+            await conn.execute(
+                f"""
+                SELECT pg_advisory_unlock({self.job_lock_keyspace}, lock_key)
+                FROM {self.jobs_table}
+                WHERE key = ANY($1)
+                """,
+                self._releasing,
+            )
+            await txn.commit()
+        self._releasing.clear()
+
+    @asynccontextmanager
+    async def _get_dequeue_conn(self) -> AsyncGenerator[PoolConnectionProxy]:
+        async with self._connection_lock:
+            if self._dequeue_conn:
+                try:
+                    await self._dequeue_conn.execute("SELECT 1")
+                except (ConnectionDoesNotExistError, InterfaceError):
+                    await self.pool.release(self._dequeue_conn)
+                    self._dequeue_conn = await self.pool.acquire()
+            else:
+                self._dequeue_conn = await self.pool.acquire()
+
+            yield self._dequeue_conn
+
+
+    @asynccontextmanager
+    async def _nullcontext(self, enter_result: ContextT) -> AsyncGenerator[ContextT]:
+        """Async version of contextlib.nullcontext
+
+        Async support has been added to contextlib.nullcontext in Python 3.10.
+        """
+        yield enter_result
+
+
+class ListenMultiplexer(Multiplexer):
+    def __init__(self, pool: Pool, key: str) -> None:
+        super().__init__()
+        self.pool = pool
+        self.key = key
+        self._connection: PoolConnectionProxy | None = None
+
+    async def _start(self) -> None:
+        if self._connection is None:
+            self._connection = await self.pool.acquire()
+        txn = self._connection.transaction()
```

Comment:
Leaving this as a context manager as it was in the `psycopg` implementation caused the connection to be released too early

Reply:
There's no way to disable `autocommit` at the driver level in `asyncpg`; creating the transaction manually.